### PR TITLE
Support "dry run" mode for updating Desired Nodes

### LINKED ISSUE #87781: Add the ability to run the Desired Nodes API in dry mode

### PR DESCRIPTION
**Pull request by arteam**
Add the `dry_run` query parameter to support simulating of updating of desired nodes. The update request will be validated, but no cluster state updates will be performed. In order to indicate that the response was a result of a dry run, we add the `dry_run` run field to the JSON representation of a response.

See #82975

### CODE REVIEW

File: docs/reference/cluster/update-desired-nodes.asciidoc [hunk 1]
```diff
@@ -92,7 +100,8 @@ The API returns the following result:
 [source,console-result]
 --------------------------------------------------
 {
-  "replaced_existing_history_id": false
+  "replaced_existing_history_id": false,
+  "dry_run": false
```
Comment (fcofdez) [inline docs/reference/cluster/update-desired-nodes.asciidoc:104]: I'm not entirely convinced about the need of returning this in the response, the caller should know that this was invoked with dry run?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/UpdateDesiredNodesRequest.java [hunk 2]
```diff
@@ -42,18 +43,24 @@ public class UpdateDesiredNodesRequest extends AcknowledgedRequest<UpdateDesired
     }
 
     public UpdateDesiredNodesRequest(String historyID, long version, List<DesiredNode> nodes) {
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/UpdateDesiredNodesRequest.java:45]: I would prefer to remove this constructor, that way we won't miss any necessary changes in tests, i.e. there's some coverage missing in `UpdateDesiredNodesRequestSerializationTests#mutateInstance`

File: server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java [hunk 3]
```diff
@@ -183,16 +183,22 @@ public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateDe
             for (final var taskContext : taskContexts) {
 
                 final var previousDesiredNodes = desiredNodes;
+                DesiredNodes newDesiredNodes;
                 try {
-                    desiredNodes = updateDesiredNodes(desiredNodes, taskContext.getTask().request());
+                    newDesiredNodes = updateDesiredNodes(desiredNodes, taskContext.getTask().request());
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:188]: nit: maybe we can extract `taskContext.getTask().request()` to a variable so the code is a bit shorter?

**Review by fcofdez — COMMENTED** [review]
Sorry for the delay on reviewing this. This looks mostly good, I left a few comments that should be tackled.

File: docs/reference/cluster/update-desired-nodes.asciidoc (hunk 1 above)
Comment (arteam) [reply docs/reference/cluster/update-desired-nodes.asciidoc:104]: I wanted add it for the consistency stake. There are a few APIs that offer the dry run option and they return the `dry_run` field.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java [hunk 4]
```diff
@@ -183,16 +183,22 @@ public ClusterState execute(ClusterState currentState, List<TaskContext<UpdateDe
...
                 } catch (Exception e) {
                     taskContext.onFailure(e);
                     continue;
                 }
+                if (taskContext.getTask().request().isDryRun() == false) {
+                    desiredNodes = newDesiredNodes;
+                }
                 final var replacedExistingHistoryId = previousDesiredNodes != null
```
[the remark this answers is not in the document]
Comment (arteam) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:201]: I've extract the dry run to a separate branch where we don't updated `desiredNodes`

File: server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java (hunk 3 above)
Comment (arteam) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java]: Done!

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

File: docs/reference/cluster/update-desired-nodes.asciidoc [hunk 5]
```diff
@@ -58,6 +62,10 @@ this API to let Elasticsearch know about the cluster topology, including future
 changes such as adding or removing nodes. Using this information, the system is
 able to take better decisions.
 
+It's possible to run the update in "dry run" mode by adding the
+`?dry_run` query parameter, or by passing `"dry_run": true` in the request
```
Comment (fcofdez) [inline docs/reference/cluster/update-desired-nodes.asciidoc:66]: > or by passing `"dry_run": true` in the request body

I think this is incorrect, we only support the URI query parameter and I would prefer to keep it that way.

**Review by fcofdez — COMMENTED** [review]
Looks good, I left a minor comment about the documentation.

File: docs/reference/cluster/update-desired-nodes.asciidoc (hunk 5 above)
Comment (arteam) [reply docs/reference/cluster/update-desired-nodes.asciidoc:66]: Good catch! To be honest, we accept it in both forms.

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/20_dry_run.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      _internal.update_desired_nodes:
+        history_id: "test"
+        version: 2
+        dry_run: "true"
+        body:
+          nodes:
+            - { settings: { "node.name": "instance-000187" }, processors: 8.5, memory: "64gb", storage: "128gb", node_version: $es_version }
+            - { settings: { "node.name": "instance-000188" }, processors: 16.0, memory: "128gb", storage: "1tb", node_version: $es_version }
+  - match: { replaced_existing_history_id: false }
+  - match: { dry_run: true }
```
Comment (fcofdez) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/20_dry_run.yml:74]: I think we miss a get_desired_nodes here to assert that the previous desired nodes are still there?

**Review by fcofdez — APPROVED** [review]
LGTM, there's an assertion missing but you can address it and merge without another review round. Thanks! 👍

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/20_dry_run.yml (hunk 6 above)
Comment (arteam) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/20_dry_run.yml:74]: Done!

**Comment by arteam** [comment]
@elasticmachine run elasticsearch-ci/part-2

**Comment by arteam** [comment]
Thank you @fcofdez!

**Pull request merged**